### DISCUSSION TITLE
Launching memtest86+ from systemd-boot

### DISCUSSION THREAD

**Discussion by mattia-b89** [memtest86plus/memtest86plus#199] [Q&A] [answered]
Hi folks,  
I'm trying to execute/launch memtest86+ from systemd-boot but with no luck.
Following tips from https://github.com/memtest86plus/memtest86plus/discussions/96 then [https://github.com/wagnerck/pop-diagnostics/blob/main/src/memtest-util.sh](https://github.com/wagnerck/pop-diagnostics/blob/main/src/memtest-util.sh) I create a new boot entry:
```
$ cat /boot/loader/entries/memtest86+.conf 
title     Memory Tester (memtest86+)
efi       /memtest86+/memtest.efi
```
while binary was installed under `/boot/memtest86+/memtest.efi` by [package](https://archlinux.org/packages/extra/any/memtest86+-efi/).  

Unfortunately when I choose the above-mentioned entry, I get a blank screen and laptop is no responsive (CTRL-ALT-CANC does nothing...) and I have to long-press the power button to power off my pc.  

**Any idea on how to get this working?**

I am on a Dell XPS 13 laptop (model: [9343](https://wiki.archlinux.org/title/Dell_XPS_13_(9343))) w/ Arch Linux

**Comment by x86fr** [comment]
Please make sure that Secureboot is disabled

**Comment by mattia-b89** [comment]
It's disabled

```
$ bootctl status
System:
     Firmware: UEFI 2.31 (American Megatrends 5.09)
  Secure Boot: disabled (disabled)
```

**Comment by debrouxl** [comment]
Does memtest86+ work on your computer when it is launched by GRUB instead of systemd-boot ?

**Reply by mattia-b89** [reply]
> Does memtest86+ work on your computer when it is launched by GRUB instead of systemd-boot ?

Yes,  
or rather: I haven't grub on my system but we can easily assume it's working, as other has pointed out the same [here](https://bugs.archlinux.org/task/76390#comment212478) and confirmed it is working with grub

**Comment by mattia-b89** [comment]
thanks to a tip from Arch, I get it loading:
if I "fix" the file with this command:
```
$ sudo objcopy memtest.efi memtest-fixed.efi
```
I am able to launch memtest86+ from systemd-boot

I am not an expert in this field and I do not know exactly how that command works, except for the result:
```
$ file /boot/memtest86+/*
/boot/memtest86+/memtest.efi:       Linux kernel x86 boot executable bzImage, version \353fHdrS\014\002, RW-rootFS,
/boot/memtest86+/memtest-fixed.efi: PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
```

now, the question is: **is it worth to include it upstream?**

**Comment by wagnerck** [comment]
Hi, I'm the guy who was working on the `pop-diagnostics` thing, with a new username. I'm no longer with System76 so I've kind of left that project alone, although I may come back to it later if `systemd-boot` becomes more widely used outside of Pop!OS.

The DEB package building process I used in `pop-diagnostics` would compile a `memtest.efi` binary from source, and it looks like the one from the Arch package is somehow being built differently. As you can see from the `file` output, it's technically not an EFI binary, even though it has an `.efi` extension, and the `objcopy` command is essentially re-writing the binary to match the extension type.

I _think_ the problem here is actually with the Arch package and not anything being distributed by the _memtest+_ team. I don't currently have an Arch system to test this with, however.

[edit] Just found [this other thread](https://github.com/memtest86plus/memtest86plus/discussions/190) which links to an official set of binaries, including the EFI file, here: https://memtest.org/download/v6.00/mt86plus_6.00.binaries.zip

**Reply by medhefgo** [reply]
The efi binaries are both a bzImage and an efi binary (it's the same layout employed by the kernel image). `file` just gives bzImage detection a higher priority I guess. The reason they're not working is explained in https://github.com/memtest86plus/memtest86plus/issues/202.

**Reply by wagnerck** [reply]
Ah, okay. That makes more sense than my theory.

**Comment by mattia-b89** [comment, marked as answer]
Thanks newest 6.10 version we can now easily launch memtest86+ from systemd-boot
